proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,226 @@
+
+Packages, as expressed by the Swift Package Manager, are a unit of code distribution. Some packages contain just a single module, but it's frequently useful to split a package's code into multiple modules. For example, when a module contains some internal helper APIs, those APIs can be split out into a utility module and maybe reused by other modules or packages.
+
+However, because Swift does not recognize organizations of code above the module level, it is not possible to create APIs like this that are purely internal to the package. To be usable from other modules within the package, the API must be public, but this means it can also be used outside of the package. This allows clients to form unwanted source dependencies on the API. It also means the built module has to export the API, which has negative implications for code size and performance.
+
+For example, here’s a scenario where a client (App) has access to a utility API from a package (gamePkg) it depends on.
+
+```
+App (Xcode project or appPkg)
+  |— Game (gamePkg)
+      |— Engine (gamePkg)
+```
+
+Here are source code examples.
+
+```
+[Engine]
+
+public struct MainEngine {
+    public init() { ... }
+    // Intended to be public
+    public var stats: String { ... }
+    // A helper function made public only to be accessed by Game
+    public func run() { ... }
+}
+
+[Game]
+
+import Engine
+
+public func play() {
+    MainEngine().run() // Can access `run` as intended since it's within the same package
+}
+
+[App]
+
+import Game
+import Engine
+
+let engine = MainEngine()
+engine.run() // Can access `run` from App even if it's not an intended behavior
+Game.play()
+print(engine.stats) // Can access `stats` as intended
+```
+
+In the above scenario, App can import Engine (a utility module in 'gamePkg') and access its helper API directly, even though the API is not intended to be used outside of its package.
+
+These costs are particularly unfortunate given that packages are a unit of code distribution. Boundaries between packages often reflect divisions between development teams; different packages are developed on different schedules by different groups of people. These boundaries are therefore especially important to enforce, because accidental dependencies can require extra work and cooperation to detangle.
+
+
+## Proposed solution
+
+Our goal is to introduce a mechanism to Swift to recognize a package as a unit in the aspect of access control. We proposed to do so by introducing a new access modifier `package`. The `package` access modifier will allow accessing symbols from outside of its defining module as long as they are within the same package. This will help set clear boundaries between packages. 
+
+This also paves the path to include a package name in the namespace in the future, which will help disambiguate modules with the same name.  
```

Comment:
```suggestion
This also paves the path to including a package name in the namespace in the future, which will help disambiguate modules with the same name.  
```

But really I'm not sure this belongs here.  It doesn't *really* pave this path, either; you can think of it as the entire utility module being package-private, but that's not going to be explicit in the source anywhere, so it doesn't actually rely on this feature existing.